src/nvim/api/command.c

```diff
@@ -779,41 +770,46 @@ static void build_cmdline_str(char **cmdlinep, exarg_T *eap, CmdParseInfo *cmdin
 
   // Keep the index of the position where command name starts, so eap->cmd can point to it.
   size_t cmdname_idx = cmdline.size;
-  kv_printf(cmdline, "%s", eap->cmd);
+  kv_concat(cmdline, eap->cmd);
 
   // Command bang.
   if (eap->argt & EX_BANG && eap->forceit) {
-    kv_printf(cmdline, "!");
+    kv_concat(cmdline, "!");
   }
 
   // Command register.
   if (eap->argt & EX_REGSTR && eap->regname) {
     kv_printf(cmdline, " %c", eap->regname);
   }
 
-  // Iterate through each argument and store the starting index and length of each argument
-  size_t *argidx = xcalloc(argc, sizeof(size_t));
   eap->argc = argc;
   eap->arglens = xcalloc(argc, sizeof(size_t));
+  size_t argstart_idx = cmdline.size;
   for (size_t i = 0; i < argc; i++) {
-    argidx[i] = cmdline.size + 1;  // add 1 to account for the space.
-    eap->arglens[i] = STRLEN(args[i]);
-    kv_printf(cmdline, " %s", args[i]);
+    String s = args.items[i].data.string;
+    eap->arglens[i] = s.size;
+    kv_concat(cmdline, " ");
+    kv_concat_len(cmdline, s.data, s.size);
   }
 
+  // Done appending to cmdline, ensure it is NUL terminated
+  kv_printf(cmdline, "");
+
   // Now that all the arguments are appended, use the command index and argument indices to set the
   // values of eap->cmd, eap->arg and eap->args.
   eap->cmd = cmdline.items + cmdname_idx;
   eap->args = xcalloc(argc, sizeof(char *));
+  size_t offset = argstart_idx;
   for (size_t i = 0; i < argc; i++) {
-    eap->args[i] = cmdline.items + argidx[i];
+    offset++;  // Account for space
+    eap->args[i] = cmdline.items + offset;
+    offset += eap->arglens[i];
   }
   // If there isn't an argument, make eap->arg point to end of cmdline.
   eap->arg = argc > 0 ? eap->args[0] : cmdline.items + cmdline.size;
```

Comment:
This points to a character *after* the null byte (invalid), instead of pointing *at* the null byte. This fixes that sanitizer error.
```suggestion
  eap->arg = argc > 0 ? eap->args[0] : cmdline.items + cmdline.size - 1;
```

Reply:
The current change doesn't get any errors.

Reply:
Okay, that's almost what's happening. Basically sometimes the null byte is included in the size of cmdline and sometimes it's not. Replace all `kv_push` with `kv_concat` to keep it consistent.

Reply:
Oh right, `kv_concat_len` could be problematic here, since it might not add a null byte either.

Reply:
I think the best way to solve this problem is just writing null at the end and backing off with the size:
```c
kv_push(cmdline, NUL);
cmdline.size--;
```
Then you should be able to do anything you want before that.

Reply:
Yes! This is the way. Thanks!

Reply:
One correction, `kv_concat` will not write the null byte either, because `strlen` doesn't include it.

Reply:
I tried something that doesn't back off the size. What do you think?

Reply:
So my initial suggestion of subtracting 1? Sure, but there probably should be a comment explaining why is it there.

Reply:
Also now because of that `kv_push(cmdline, NUL)` should be done unconditionally.